lib/mlton/basic/hash-table.sml

```diff
@@ -1,104 +1,56 @@
-(* Copyright (C) 1999-2006, 2008 Henry Cejtin, Matthew Fluet, Suresh
+(* Copyright (C) 2018 Jason Carr
+ * Copyright (C) 1999-2006, 2008 Henry Cejtin, Matthew Fluet, Suresh
  *    Jagannathan, and Stephen Weeks.
  *
  * MLton is released under a HPND-style license.
  * See the file MLton-LICENSE for details.
  *)
-
-(* This code is not working -- it is not even in sources.cm *)
 structure HashTable: HASH_TABLE =
 struct
 
 structure Set = HashSet
 
-type ('a, 'b) t = ('a * 'b) Set.t
+datatype ('a, 'b) t = T of {set: {hash: word, key: 'a, value: 'b} Set.t,
+                        hash: 'a -> word,
+                        equals: 'a * 'a -> bool}
```

Comment:
If you make `equals : 'a -> 'a -> bool` and perform the currying in `new`, then you won't need `keyEquals`.